design/design-doc-developer-experience-tooling.md

```diff
@@ -0,0 +1,498 @@
+varies by language.
+
+#### Extensibility
+
+Similar to function builders, schema generation is designed to be extensible. A
+schema generator implementation is given a directory of CRDs and outputs a
+directory of language bindings. The output directory can have an arbitrary
+layout, since every language expects files to be organized differently.
+
+### Testing
+
+Testing is another key facet of software development facilitated by
+projects. Projects allow for three layers of testing:
+
+1. Language-specific tests for embedded functions (e.g., Go or Python unit
+   tests).
+2. Composition tests, which use `crossplane render` to run compositions
+   (including embedded functions) and check assertions on the output.
+3. E2E tests, which install a project's packages into a real control plane,
+   apply resources, and wait for them to have certain conditions. E2E tests are
+   executed using [uptest].
+
+Language-specific tests may use the language bindings described above, but
+otherwise are built using language-specific tools outside the scope of this
+design.
+
+Composition tests and E2E tests are written as YAML manifests describing the
+test to run. The tooling will include the ability to generate test manifests
+from code on-the-fly (in the same languages supported for embedded functions),
+so that extensive test suites can be built easily without duplicating many lines
+of YAML. Multiple tests can be specified at once, and will be run in sequence.
+
+The composition test API looks like this:
+
+```yaml
+apiVersion: dev.crossplane.io/v1alpha1
+kind: CompositionTest
+metadata:
+  name: test-cluster
+spec:
+  # The XR to render as input to the test.
+  xrPath: examples/cluster/xr.yaml
+  # The XRD, for schema validation.
+  xrdPath: apis/cluster/definition.yaml
+  # The composition to execute for the test.
+  compositionPath: apis/cluster/composition.yaml
+  # Optional observed resources for the composition pipeline, e.g. to test
+  # conditional logic.
+  observedResources: []
+  # Assertions on the resources rendered by the test, which can include any
+  # expected updates to the XR as well as composed resources.
+  assertions:
+    # Use chainsaw to compare resources.
+    - type: chainsaw
+      chainsaw:
+        resources:
+          - apiVersion: platform.example.com/v1alpha1
+            kind: Cluster
+            metadata:
+              name: example
+            spec:
+              version: 1.33
+              region: us-west1
+          - apiVersion: container.gcp.upbound.io/v1beta1
+            kind: Cluster
+            metadata:
+              annotations:
+                crossplane.io/composition-resource-name: cluster
+            spec:
+              forProvider:
+                location: us-west1
+                minMasterVersion: 1.33
+                nodeVersion: 1.33
+  # Timeout for the test.
+  timeoutSeconds: 120
+  # Whether to validate the output of the render.
+  validate: false
```

Comment:
In xprin the `crossplane beta validate` is run only if an array of CRDs (or crossplane.yaml) is specified